File: proj/CoreLibrary/src/com/jasonzqshen/familyaccounting/core/reports/DocumentBusinessIndex.java
package com.jasonzqshen.familyaccounting.core.reports;

import com.jasonzqshen.familyaccounting.core.CoreDriver;
import com.jasonzqshen.familyaccounting.core.masterdata.MasterDataIdentity;
import com.jasonzqshen.familyaccounting.core.masterdata.MasterDataManagement;
import com.jasonzqshen.familyaccounting.core.masterdata.MasterDataType;
import com.jasonzqshen.familyaccounting.core.transaction.HeadEntity;
import com.jasonzqshen.familyaccounting.core.transaction.ItemEntity;
import com.jasonzqshen.familyaccounting.core.utils.CreditDebitIndicator;
import com.jasonzqshen.familyaccounting.core.utils.CurrencyAmount;

/**
 * document index based on business area
 * 
 * @author Jason
 * 
 */
public class DocumentBusinessIndex extends DocumentIndex {

    protected DocumentBusinessIndex(CoreDriver coreDriver,
            MasterDataManagement mdMgmt) {
        super(coreDriver, mdMgmt);
    }

    @Override
    protected void newDoc(HeadEntity head) {
        ItemEntity[] items = head.getItems();
        for (int i = 0; i < items.length; ++i) {
            MasterDataIdentity id = items[i].getBusinessArea();

            if (id != null) {
                DocumentIndexItem item;
                if (!_list.containsKey(id)) {
                    item = new DocumentIndexItemWithBalance(_coreDriver, id,
                            MasterDataType.BUSINESS_AREA);
                    _list.put(id, item);
                } else {
                    item = _list.get(id);
                }
                // add document
                item.addDoc(head);
                // add amount
                CurrencyAmount amount = items[i].getAmount();
                if (items[i].getCDIndicator() == CreditDebitIndicator.CREDIT) {
                    amount.negate();
                }
                item.addAmount(head.getMonthId(), amount);
            }
        }
    }

    @Override
    public DocumentIndexItemWithBalance getIndexItem(MasterDataIdentity key) {
        return (DocumentIndexItemWithBalance) super.getIndexItem(key);
    }
}

File: trunk/proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingCore/Reports/DocumentBusinessIndex.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using WinStore_FamilyAccountingCore.MasterData;
using WinStore_FamilyAccountingCore.Transaction;
using WinStore_FamilyAccountingCore.Utilities;

namespace WinStore_FamilyAccountingCore.Reports
{
    public class DocumentBusinessIndex : DocumentIndex
    {
        internal DocumentBusinessIndex(CoreDriver coreDriver,
                MasterDataManagement mdMgmt)
            : base(coreDriver, mdMgmt)
        {
        }

        protected override void newDoc(HeadEntity head)
        {
            List<ItemEntity> items = head.Items;
            foreach (ItemEntity itemEntity in items)
            {
                MasterDataIdentity id = itemEntity.GLAccount;
                if (id != null)
                {
                    DocumentIndexItem item;
                    if (!_list.TryGetValue(id, out item))
                    {
                        item = new DocumentIndexItemWithBalance(_coreDriver, id,
                                MasterDataType.BUSINESS_AREA);
                        _list.Add(id, item);
                    }
                    // add document
                    item.addDoc(head);
                    // add amount
                    CurrencyAmount amount = itemEntity.Amount;
                    if (itemEntity.CdIndicator == CreditDebitIndicator.CREDIT)
                    {
                        amount.Negate();
                    }
                    item.addAmount(head.MonthID, amount);
                }
            }
        }

        /// <summary>
        /// reverse document
        /// </summary>
        /// <param name="head"></param>
        protected override void reverseDoc(HeadEntity head)
        {
            List<ItemEntity> items = head.Items;
            foreach (ItemEntity itemEntity in items)
            {
                MasterDataIdentity id = itemEntity.BusinessArea;
                if (id == null)
                {
                    continue;
                }
                DocumentIndexItem item;
                if (!_list.TryGetValue(id, out item))
                {
                    continue;
                }
                // remove document
                item.removeDoc(head);
                // reverse amount
                CurrencyAmount amount = itemEntity.Amount;
                if (itemEntity.CdIndicator == CreditDebitIndicator.DEBIT)
                {
                    amount.Negate();
                }
                item.addAmount(head.MonthID, amount);
            }

        }
    }

}
